az-cvm-vtpm/az-tdx-vtpm/src/hcl.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use crate::tdx::TdxVmReport;
+use jsonwebkey::JsonWebKey;
+use memoffset::offset_of;
+use serde::{Deserialize, Serialize};
+use serde_big_array::BigArray;
+use sev::firmware::guest::AttestationReport as SnpVmReport;
+use sha2::{Digest, Sha256};
+use std::mem;
+use thiserror::Error;
+
+const HCL_AKPUB_KEY_ID: &str = "HCLAkPub";
+const MAX_REPORT_SIZE: usize = mem::size_of::<SnpVmReport>();
+const MIN_REPORT_SIZE: usize = mem::size_of::<TdxVmReport>();
+
+#[derive(Error, Debug)]
+pub enum HclError {
+    #[error("invalid report type")]
+    InvalidReportType,
+    #[error("AkPub not found")]
+    AkPubNotFound,
+    #[error("parse error")]
+    ParseError(#[from] bincode::Error),
+}
+
+#[derive(Deserialize, Debug)]
+struct VarDataKeys {
+    keys: Vec<JsonWebKey>,
+}
+
+#[repr(u32)]
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+enum IgvmHashType {
+    Invalid = 0,
+    Sha256,
+    Sha384,
+    Sha512,
+}
+
+#[repr(u32)]
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+enum IgvmReportType {
+    Invalid = 0,
+    VbsVm,
+    SnpVm,
+    Tvm,
+    TdxVm,
+}
```

Comment:
It is interesting that this field is different in SNP CVMs vs TDX CVMs.

Reply:
hmm, it's a enum that's extended and what used to be reserved is now called "vbsvm". I wonder whether we should limit this enum to to Snp + Tdx, so parsing will fail for everything else.

The downside: parsing errors could be very obscure (unexpected: 3)